lexicon/providers/hetzner.py

```diff
@@ -0,0 +1,604 @@
+                                 name, max_concats)
+                    self._exit_session()
+                    raise AssertionError
+                qname = cname if cname else name
+                domain, nameservers = self._get_dns(domain, qname)
+                rrset = self._dns_lookup(qname, 'CNAME')
+                if rrset:
+                    concat += 1
+                    cname = rrset[0].to_text()
+                else:
+                    concatenate = False
+        LOGGER.info('Hetzner => Record %s has CNAME %s', name, cname)
+        return domain, nameservers, cname
+
+    # Hetzner Helpers
+    ###########################################################################
+    @staticmethod
+    def _extract_domain_id(string, regex):
+        regex = re.compile(regex)
+        match = regex.search(string)
+        if not match:
+            return False
+        return str(match.group(1))
+
+    @staticmethod
+    def _extract_hidden_data(dom):
+        input_tags = dom.find_all('input', attrs={'type': 'hidden'})
+        data = {}
+        for input_tag in input_tags:
+            data[input_tag['name']] = input_tag['value']
+        return data
+
+    @staticmethod
+    def _filter_dom(dom, filters, last_find_all=False):
+        if isinstance(dom, six.string_types):
+            dom = BeautifulSoup(dom, 'html.parser')
+        for idx, find in enumerate(filters, start=1):
+            if not dom:
+                break
+            name = find.get('name', None)
+            attrs = find.get('attrs', {})
+            if len(filters) == idx and last_find_all:
+                dom = dom.find_all(name, attrs=attrs) if name else dom.find_all(attrs=attrs)
+            else:
+                dom = dom.find(name, attrs=attrs) if name else dom.find(attrs=attrs)
+        return dom
+
+    def _auth_session(self, username, password):
+        api = self.api[self.account]['auth']
+        endpoint = api.get('endpoint', self.api[self.account]['endpoint'])
+        session = requests.session()
+        response = session.request('GET', endpoint + api['GET'].get('url', '/'))
+        dom = Provider._filter_dom(response.text, api['filter'])
+        data = Provider._extract_hidden_data(dom)
+        data[api['user']], data[api['pass']] = username, password
+        response = session.request('POST', endpoint + api['POST']['url'], data=data)
+        if Provider._filter_dom(response.text, api['filter']):
+            LOGGER.error('Hetzner => Unable to authenticate session with %s account \'%s\': '
+                         'Invalid credentials',
+                         self.account, username)
+            raise AssertionError
+        LOGGER.info('Hetzner => Authenticate session with %s account \'%s\'',
+                    self.account, username)
+        return session
+
+    def _exit_session(self):
+        if self._get_provider_option('live_tests') is None:
```

Comment:
Code is aware to be executed in a test. See my comment for this topic below.

Reply:
I do not know how to change this. At this point I want to avoid that the test script exit the session because all action calls will be done with only one authenticate call and the second time I use the live_tests env var it fasten up the offline test for Hetzner Robot.
There is no given option to manipulate, because only the abnormal way the actions are executed in the test is the reason the script fails. I would need to create one anyway, ar'nt I?

Reply:
Your solution overcome the problem for the tests, but you will still have it outside tests. Because lexicon is used as a library, not only as a cli. And as a library, a developer would expect to create a lexicon client, call the authenticate method, and then execute several actions. So it will fail, because the session will be claused after the first action.

I think it is a design problem. You want to open the session during the authentication phase. You should open the session for each action then close at the end. Like this, each action is independant from each other, and all other problems that are related to the ending session handling that I noted would also vanish. 

The question now is: why do you want to create the session during the authentication phase, and is it a problem to have it open and closed several time if there is several actions (like for instance during a create/update/delete that will need also to call a list at some point).

If it is only a problem of save some requests,I will say this to you: I also tried to do it with ovh, it gave me only problems. Better to have some requests more, less speed, and have a complete isolation between actions.

Reply:
Yeaahhh, you nailed it! ^^ I changed it. After that I think there is only one point left. How to fasten up offline test. Now on every create|update|delete call who changes the zone file on Hetzner Robot, there will be a delay of 30s, because in live mode Hetzner need this time to took over the updated zone file.

Reply:
I add a similar problem with godaddy, because the API does not propagate immediately the changes. The zone was in a pending state, and you should poll until the API respond a correct status. So during the replay, requests retry capability was also triggered by the recorded responses that would return multiple time the pending status before responding a valid status.

Well, I solved this manually. After recording, and before merging, I just remove almost all pending status response, except on or two for a few tests, to ensure the retry logic is still triggered.

Reply:
That won't work here. The API responses with status code 200 and returns the unchanged zone file version.

I could look repeatedly until the changes are took over, but that came with a big overload... Any other idea?

Reply:
Yes, I added another proposition below.